src/mqtt/transport/tcp/mqtt_tcp.c

```diff
@@ -1137,13 +1137,15 @@ mqtt_tcptran_ep_fini(void *arg)
 
 	nni_mtx_lock(&ep->mtx);
 	ep->fini = true;
-	if (ep->connmsg)
-		nni_msg_free(ep->connmsg);
 	if (ep->refcnt != 0) {
 		nni_mtx_unlock(&ep->mtx);
 		return;
 	}
 	nni_mtx_unlock(&ep->mtx);
+	// Free connmsg once
+	if (ep->connmsg)
+		nni_msg_free(ep->connmsg);
+
```

Comment:
Maybe do this in `ep->mtx` is better?

Reply:
I think the refcnt has ensured that the free only executes once.